staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,247 @@ limitations under the License.
-				continue
-			}
-			return nil, err
-		}
-		if delegateSpec == nil {
+		// ignore errors for the empty delegate we attach at the end the chain
+		// atm the empty delegate returns 503 when the server hasn't been fully initialized
+		// and the spec downloader only silences 404s
+		if len(delegate.ListedPaths()) == 0 && delegate.NextDelegate() == nil {
 			continue
 		}
-		s.addLocalSpec(delegateSpec, handler, fmt.Sprintf(localDelegateChainNamePattern, i), etag)
-		i++
+		delegationList = append(delegationList, handler)
 	}
-
-	// Build initial spec to serve.
-	klog.V(2).Infof("Building initial OpenAPI spec")
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished initial OpenAPI spec generation after %v", duration)
-
-		regenerationCounter.With(map[string]string{"apiservice": "*", "reason": "startup"})
-		regenerationDurationGauge.With(map[string]string{"reason": "startup"}).Set(duration.Seconds())
-	}(time.Now())
-	specToServe, err := s.buildOpenAPISpec()
-	if err != nil {
-		return nil, err
-	}
-
-	// Install handler
-	s.openAPIVersionedService = handler.NewOpenAPIService(specToServe)
-	s.openAPIVersionedService.RegisterOpenAPIVersionedService("/openapi/v2", pathHandler)
-
+	s := buildAndRegisterSpecAggregatorForLocalServices(downloader, aggregatorOpenAPISpec, delegationList, pathHandler)
 	return s, nil
 }
 
-type specAggregator struct {
-	// mutex protects all members of this struct.
-	rwMutex sync.RWMutex
-
-	// Map of API Services' OpenAPI specs by their name
-	openAPISpecs map[string]*openAPISpecInfo
-
-	// provided for dynamic OpenAPI spec
-	openAPIVersionedService *handler.OpenAPIService
-}
-
-var _ SpecAggregator = &specAggregator{}
-
-// This function is not thread safe as it only being called on startup.
-func (s *specAggregator) addLocalSpec(spec *spec.Swagger, localHandler http.Handler, name, etag string) {
-	localAPIService := v1.APIService{}
-	localAPIService.Name = name
-	s.openAPISpecs[name] = &openAPISpecInfo{
-		etag:       etag,
-		apiService: localAPIService,
-		handler:    localHandler,
-		spec:       spec,
+func (s *specAggregator) addLocalSpec(name string, spec cached.Data[*spec.Swagger]) {
+	service := v1.APIService{}
+	service.Name = name
+	info := &openAPISpecInfo{
+		apiService: service,
 	}
+	info.spec.Replace(spec)
+	s.openAPISpecs[name] = info
 }
 
-// openAPISpecInfo is used to store OpenAPI spec with its priority.
-// It can be used to sort specs with their priorities.
-type openAPISpecInfo struct {
-	apiService v1.APIService
-
-	// Specification of this API Service. If null then the spec is not loaded yet.
-	spec    *spec.Swagger
-	handler http.Handler
-	etag    string
-}
+// buildMergeSpecLocked creates a new cached mergeSpec from the list of cached specs.
+func (s *specAggregator) buildMergeSpecLocked() cached.Data[*spec.Swagger] {
+	apiServices := make([]*v1.APIService, 0, len(s.openAPISpecs))
+	caches := make([]cached.Data[*spec.Swagger], len(s.openAPISpecs))
```

Comment:
defining vars at the top is an anti-pattern in Go. Move them directly above their loop.